src/runtime/refiner.ts

```diff
@@ -0,0 +1,158 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Refinement, SchemaPrimitiveType, RefinementExpression} from './manifest-ast-nodes.js';
+import {Dictionary} from './hot.js';
+import {assert} from '../platform/assert-node.js';
+import {Entity} from './entity.js';
+import {Schema} from './schema.js';
+
+export class Refiner {
+    // Converts refinement ast-node to string.
+    // Ast-nodes are typically interfaces with no behaviour.
+    static refinementString(refinement: Refinement): string {
+        if (!refinement) {
+            return '';
+        }
+        return '[' + Refiner.expressionString(refinement.expression) + ']';
+    }
+
+    private static expressionString(expr: RefinementExpression): string {
+        if (expr.kind === 'binary-expression-node') {
+            return '(' + Refiner.expressionString(expr.leftExpr) + ' ' + expr.operator + ' ' + Refiner.expressionString(expr.rightExpr) + ')';
```

Comment:
It would be nice to have the toString behavior on the expression itself.
Later we can convert from ast nodes to a runtime representation (not that it will change much, but it should let us inject the toString an eval functions).